### PR TITLE
[Test] connect to local @open sesame 07/20 14:28

### PR DESCRIPTION
**Pull request by jaeyun-jung**
Add testcase for edge connection and data transfer between local server and client.

Signed-off-by: Jaeyun <jy1210.jung@samsung.com>

### CODE REVIEW

File: src/libnnstreamer-edge/nnstreamer-edge-internal.c [hunk 1]
```diff
@@ -1287,6 +1288,8 @@ nns_edge_request (nns_edge_h edge_h, nns_edge_data_h data_h)
   }
 
   ret = _nns_edge_cmd_send (conn_data->sink_conn, &cmd);
+  if (ret != NNS_EDGE_ERROR_NONE)
+    nns_edge_loge ("Failed to request, cannot send edge data.");
```
Comment (myungjoo) [inline src/libnnstreamer-edge/nnstreamer-edge-internal.c:1292]: Let's print out the error code or its reason, too.

File: tests/unittest_nnstreamer-edge.cc [hunk 2]
```diff
@@ -13,32 +13,106 @@
...
+  if (!_td)
+    return;
+
+  if (_td->loop) {
+    if (g_main_loop_is_running (_td->loop))
+      g_main_loop_quit (_td->loop);
+
+    g_main_loop_unref (_td->loop);
+    _td->loop = NULL;
+  }
```
Comment (gichan-jang) [inline tests/unittest_nnstreamer-edge.cc:60]: Let's free `_td` here.

Comment (jaeyun-jung) [reply tests/unittest_nnstreamer-edge.cc:60]: updated

File: src/libnnstreamer-edge/nnstreamer-edge-common.c [hunk 3]
```diff
@@ -15,9 +15,19 @@
 
 #include "nnstreamer-edge-common.h"
 
+/**
+ * @brief Free allocated memory.
+ */
+void
+nns_edge_free (void *data)
```
Comment (wooksong) [inline src/libnnstreamer-edge/nnstreamer-edge-common.c:22]: There is a macro, ```SAFE_FREE```. Do you have any reason to define a new function?

Comment (jaeyun-jung) [reply src/libnnstreamer-edge/nnstreamer-edge-common.c]: Related PR is https://github.com/nnstreamer/nnstreamer-edge/pull/32, this is for the release callback.

**Comment by jaeyun-jung** [comment]
To reviewers, fixed exception case with GC, and PR updated for testing the data transfer.

TODO: I will add new pull request later - adding util function to get the available port.

File: tests/unittest_nnstreamer-edge.cc [hunk 4]
```diff
@@ -47,6 +123,143 @@ _test_edge_event_cb (nns_edge_event_h event_h, void *user_data)
...
+    ret = nns_edge_request (client2_h, data_h);
+    EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+    usleep (100000);
+  }
+
+  ret = nns_edge_data_destroy (data_h);
+  EXPECT_EQ (ret, NNS_EDGE_ERROR_NONE);
+
+  sleep (2);
```
Comment (myungjoo) [inline tests/unittest_nnstreamer-edge.cc:242]: it has always been troubling when we assume that the CI/build server will be reasonably fast....

Comment (jaeyun-jung) [reply tests/unittest_nnstreamer-edge.cc]: Updated - wait for responding from server handle.

File: tests/unittest_nnstreamer-edge.cc [hunk 5]
```diff
@@ -47,6 +123,150 @@ _test_edge_event_cb (nns_edge_event_h event_h, void *user_data)
...
+  _td_client1->handle = client1_h;
+
+  nns_edge_create_handle ("temp-client2", "temp-topic", &client2_h);
+  nns_edge_set_event_callback (client2_h, _test_edge_event_cb, _td_client2);
+  nns_edge_set_info (client2_h, "CAPS", "test client2");
+  _td_client2->handle = client2_h;
+
+  /* Start server/client thread */
+  pthread_attr_init (&attr);
+  pthread_attr_setdetachstate (&attr, PTHREAD_CREATE_DETACHED);
```
Comment (gichan-jang) [inline tests/unittest_nnstreamer-edge.cc:188]: Set the thread attribute as JOINABLE?

Comment (jaeyun-jung) [reply tests/unittest_nnstreamer-edge.cc:188]: IMO, unnecessary. this thread handles test data to quit g-main-loop.

**Review by gichan-jang — APPROVED** [review]

**Pull request merged**